paasta_tools/contrib/check_orphans.py

```diff
@@ -0,0 +1,213 @@
+        if host_config is None:
+            not_found_hosts.add(host)
+        else:
+            instance_set |= read_one_nerve_file(host_config)
+    return instance_set, not_found_hosts
+
+
+def get_instance_data(service_blacklist_path: str) -> Tuple[Set[InstanceTuple], Set[InstanceTuple]]:
+    blacklisted_services = set(yaml.load(service_blacklist_path))
+    # Dump ZK
+    zk_data = get_zk_data(blacklisted_services)
+    zk_instance_data = read_from_zk_data(zk_data)
+
+    hosts = {x[0] for x in zk_instance_data}
+
+    # Dump Nerve configs from each host via xinetd
+    coro = gather_files(hosts)
+    results = asyncio.get_event_loop().run_until_complete(coro)
+
+    nerve_instance_data, not_found_hosts = read_nerve_files(results)
+
+    # Filter out anything that we couldn't get a nerve config for
+    zk_instance_data_filtered = {x for x in zk_instance_data if x[0] not in not_found_hosts}
+
+    logging.info('zk_instance_data (unfiltered) len: {}'.format(len(zk_instance_data)))
+    logging.info('zk_instance_data (filtered) len: {}'.format(len(zk_instance_data_filtered)))
+    logging.info('nerve_instance_data len: {}'.format(len(nerve_instance_data)))
+
+    return zk_instance_data_filtered, nerve_instance_data
+
+
+def write_output(
+    output_dir: str,
+    zk_instance_data: Set[InstanceTuple],
+    nerve_instance_data: Set[InstanceTuple],
+) -> None:
+    orphans = zk_instance_data - nerve_instance_data
+    if orphans:
+        logging.warning('{} orphans found'.format(len(orphans)))
+
+    pathlib.Path(output_dir).mkdir(exist_ok=True)
+
+    # groupby host
+
+    orphans_by_host: DefaultDict[str, List[Tuple[int, str]]] = defaultdict(list)
+    for orphan in orphans:
+        orphans_by_host[orphan.host].append((orphan.port, orphan.service))
+    with open(os.path.join(output_dir, 'orphans_by_host.yaml'), 'w') as f:
+        yaml.dump(dict(orphans_by_host), f)
+
+    # groupby service
+    orphans_by_service: DefaultDict[str, List[Tuple[str, int]]] = defaultdict(list)
+    for orphan in orphans:
+        orphans_by_service[orphan.service].append((orphan.host, orphan.port))
+    with open(os.path.join(output_dir, 'orphans_by_service.yaml'), 'w') as f:
+        yaml.dump(dict(orphans_by_service), f)
+
+    # collisions
+    instance_by_addr: DefaultDict[Tuple[str, int], Set[str]] = defaultdict(set)
+    for nerve_inst in nerve_instance_data:
+        instance_by_addr[(nerve_inst.host, nerve_inst.port)].add(nerve_inst.service)
+
+    collisions: List[str] = []
+    for zk_inst in zk_instance_data:
+        nerve_services = instance_by_addr[(zk_inst.host, zk_inst.port)]
+        if len(nerve_services) >= 1 and zk_inst.service not in nerve_services:
+            collisions.append(
+                f'[{zk_inst.host}:{zk_inst.port}] {zk_inst.service} collides with {nerve_services}',
+            )
+
+    if collisions:
+        logging.warning('Collisions found! Traffic is being misrouted!')
```

Comment:
Gotta set the return value, to 2 here.
Then maybe filter down the return value and do `sys.exit(main()` or something like that

Reply:
We want the check to fail when there are orphans (timebombs), not just collisions (detonated timebombs). How about I return `3` (to avoid [reserved](https://www.tldp.org/LDP/abs/html/exitcodes.html#EXITCODESREF) exit codes) for collisions and 4 for orphans?

Reply:
Heh, I wouldn't worry about the reserved return codes and instead follow the standard nagios check api
https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/pluginapi.html

0 (ok) for nothing
1 (warning) for timebombs
2 (critical) for collisions